### PR TITLE
[fuzz] Http path utility fuzzer

### PR DESCRIPTION
**Pull request by zasweq**
Commit Message:  Added 92% line Fuzz coverage to common/http/path_utility.cc
Risk Level: Low
Testing: N/A
Docs Changes: N/A

Signed-off-by: Zach Reyes <zasweq@google.com>

### CODE REVIEW

File: test/common/http/path_utility_fuzz_test.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  case test::common::http::PathUtilityTestCase::kMergeSlashes: {
+    auto request_headers = fromHeaders<Http::TestRequestHeaderMapImpl>(
+        input.merge_slashes().request_headers(), {}, {":path"});
+    Http::PathUtil::mergeSlashes(request_headers);
+    break;
+  }
+  case test::common::http::PathUtilityTestCase::kRemoveQueryAndFragment: {
+    auto path = input.remove_query_and_fragment().path();
+    Http::PathUtil::removeQueryAndFragment(path);
```
Comment (asraa) [inline test/common/http/path_utility_fuzz_test.cc:27]: could a validation be done here that the result should be a substring of the original path?

File: test/common/http/path_utility_fuzz_test.cc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+namespace Envoy {
+namespace Fuzz {
+namespace {
+DEFINE_PROTO_FUZZER(const test::common::http::PathUtilityTestCase& input) {
+  switch (input.path_utility_selector_case()) {
+  case test::common::http::PathUtilityTestCase::kCanonicalPath: {
+    auto request_headers = fromHeaders<Http::TestRequestHeaderMapImpl>(
+        input.canonical_path().request_headers(), {},
+        {":path"}); // needs to have path header in order to be valid
+    Http::PathUtil::canonicalPath(request_headers);
```
Comment (asraa) [inline test/common/http/path_utility_fuzz_test.cc:16]: is there any other validation that we can do on this? possible just verifying that the header map contains a non-empty path header string after applying this function?

**Comment by zasweq** [comment]
Yes, 4096 executions/second!

File: test/common/http/path_utility_fuzz_test.cc (hunk 2 above)
Comment (zasweq) [reply test/common/http/path_utility_fuzz_test.cc]: This fails quite a bit, simply due to the random nature of the fuzzing engine. Should I do a release assert here? I had a simple EXPECT_TRUE from the returned boolean, and also an EXPECT_NE to an empty string.

Comment (asraa) [reply test/common/http/path_utility_fuzz_test.cc:16]: Validating the return value will fail, like you said, but validating that the function preserves a path header like you have will always work

File: test/common/http/path_utility_fuzz_test.cc [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    break;
+  }
+  default:
+    break;
+  }
+}
+
+} // namespace
+} // namespace Fuzz
+} // namespace Envoy
```
Comment (asraa) [inline test/common/http/path_utility_fuzz_test.cc:39]: could you add newlines to the end of files? i think vscode has this option `"files.insertFinalNewline": true.`

File: test/common/http/path_utility_fuzz_test.cc [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+namespace Fuzz {
+namespace {
+DEFINE_PROTO_FUZZER(const test::common::http::PathUtilityTestCase& input) {
+  switch (input.path_utility_selector_case()) {
+  case test::common::http::PathUtilityTestCase::kCanonicalPath: {
+    auto request_headers = fromHeaders<Http::TestRequestHeaderMapImpl>(
+        input.canonical_path().request_headers(), {},
+        {":path"}); // needs to have path header in order to be valid
+    Http::PathUtil::canonicalPath(request_headers);
+    ASSERT(request_headers.get_(":path") != "");
```
Comment (asraa) [inline test/common/http/path_utility_fuzz_test.cc:17]: could try `getPathValue().empty()`. the `get##Name` accessor is O(1), the `get_` will have some more overhead to test whether it is an O(1) header first.

**Pull request merged**